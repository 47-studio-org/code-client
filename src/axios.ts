import axios, { AxiosRequestConfig, AxiosError, AxiosResponse } from 'axios';

const axios_ = axios.create({
  responseType: 'json',
  headers: {
    'Content-Type': 'application/json;charset=utf-8',
  },
});

axios_.interceptors.request.use(
  (config: AxiosRequestConfig) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const { method, url, data } = config;
    console.log(`=> HTTP ${method?.toUpperCase()} ${url} ${data ? JSON.stringify(data) : ''}`.slice(0, 399));

    return config;
  },
  (error: AxiosError) => {
    console.error(`Request error --> ${error.message}`);
    throw error;
  },
);

axios_.interceptors.response.use(
  (response: AxiosResponse) => {
    console.log(`<= Response: ${response.status} ${JSON.stringify(response.data)}`.slice(0, 399));
    return response;
  },
  (error: AxiosError) => {
    console.warn(`Response error --> ${error.message}`);
    throw error;
  },
);

export default axios_;
