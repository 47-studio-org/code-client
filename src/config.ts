export const PACKAGE_NAME = 'DeepCodeAI';
export const PACKAGE_SHORT_NAME = 'DC-API';
export const BASE_URL = 'https://www.deepcode.ai';
export const API_URL = '/publicapi';
