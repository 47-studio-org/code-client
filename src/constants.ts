import { CustomDCIgnore, DefaultDCIgnore } from '@deepcode/dcignore';

export const MAX_PAYLOAD = 4 * 1024 * 1024;
export const defaultBaseURL = 'https://www.deepcode.ai';
export const apiPath = '/publicapi';
export const HASH_ALGORITHM = 'sha256';
export const ENCODE_TYPE = 'hex';
export const GIT_FILENAME = '.git';
export const GITIGNORE_FILENAME = '.gitignore';
export const DCIGNORE_FILENAME = '.dcignore';
export const EXCLUDED_NAMES = [GIT_FILENAME, GITIGNORE_FILENAME, DCIGNORE_FILENAME];

export const IGNORES_DEFAULT = [`**/${GIT_FILENAME}`];

export const IGNORE_FILES_NAMES = [GITIGNORE_FILENAME, DCIGNORE_FILENAME];

export const DCIGNORE_DRAFTS = {
  custom: CustomDCIgnore,
  default: DefaultDCIgnore,
};

// // eslint-disable-next-line no-shadow
// export enum FILE_CURRENT_STATUS {
//   modified = 'modified',
//   deleted = 'deleted',
//   same = 'same',
//   created = 'created',
// }

// eslint-disable-next-line no-shadow
export enum RequestTypes {
  startSession = 'startSession',
  checkSession = 'checkSession',
  getFilters = 'getFilters',
  createBundle = 'createBundle',
  checkBundle = 'checkBundle',
  extendBundle = 'extendBundle',
  uploadFiles = 'uploadFiles',
  getAnalysis = 'getAnalysis',
  reportError = 'reportError',
  reportEvent = 'reportEvent',
}

export const ERROR_CODES = new Set([
  304, // loginInProgress,
  400, // unauthorizedContent,
  401, // unauthorizedUser,
  403, // unauthorizedBundleAccess,
  404, // notFound,
  413, // bigPayload,
  500, // serverError,
  502, // badGateway,
  503, // serviceUnavailable,
  504, // timeout,
]);

export const ERRORS = {
  [RequestTypes.startSession]: {
    other: 'Login failed',
  },
  [RequestTypes.checkSession]: {
    401: 'Missing or invalid sessionToken',
    other: 'Checking session failed',
  },
  [RequestTypes.getFilters]: {
    401: 'Missing sessionToken or incomplete login process',
    other: 'Getting filters failed',
  },
  [RequestTypes.createBundle]: {
    400: "Request content doesn't match the specifications",
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested repository',
    404: 'Unable to resolve requested oid',
    413: 'Payload too large',
    other: 'Creating bundle failed',
  },
  [RequestTypes.checkBundle]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    404: 'Uploaded bundle has expired',
    other: 'Checking bundle failed',
  },
  [RequestTypes.extendBundle]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Extending bundle failed',
  },
  [RequestTypes.uploadFiles]: {
    400: 'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to parent bundle',
    404: 'Parent bundle has expired',
    413: 'Payload too large',
    other: 'Uploading files failed',
  },
  [RequestTypes.getAnalysis]: {
    401: 'Missing sessionToken or incomplete login process',
    403: 'Unauthorized access to requested bundle',
    other: 'Getting analysis failed',
  },
  [RequestTypes.reportError]: {
    other: 'Reporting error failed',
  },
  [RequestTypes.reportEvent]: {
    other: 'Reporting event failed',
  },
};

export const BUNDLE_ERRORS = {
  create: {
    400: 'Creating bundle: Request content does not match the specifications',
    401: 'Creating bundle: Missing sessionToken or incomplete login process',
    403: 'Creating bundle: Unauthorized access to requested repository',
    413: 'Creating bundle: Payload too large',
  },
  upload: {
    400: 'Uploading files: Content and hash mismatch or attempted to upload files to a git bundle',
    401: 'Uploading files: Missing sessionToken or incomplete login process',
    403: 'Uploading files: Unauthorized access to requested bundle',
    413: 'Uploading files: Payload too large',
  },
  check: {
    401: 'Check bundle: Missing sessionToken or incomplete login process',
    403: 'Check bundle: Unauthorized access to requested bundle',
    404: 'Check bundle: Uploaded bundle has expired',
  },
  extend: {
    400: 'Extending bundle: Attempted to extend a git bundle, or ended up with an empty bundle after the extension',
    401: 'Extending bundle: Missing sessionToken or incomplete login process',
    403: 'Extending bundle: Unauthorized access to parent bundle',
    404: 'Extending bundle: Parent bundle has expired',
    413: 'Extending bundle: Payload too large',
  },
};
