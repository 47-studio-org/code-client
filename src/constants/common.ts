import path from 'path';

export const isWindows = path.sep === '\\';

// Max payload is 5 MB, but we will use 4 MB for files due to we need also to send
// another info about files with their contents
export const maxPayload = 4 * 1024 * 1024;

export const defaultBaseURL = 'https://www.deepcode.ai';
export const apiPath = '/publicapi';
