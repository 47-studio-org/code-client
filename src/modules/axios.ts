import axios from 'axios';

const axios_ = axios.create({
  responseType: 'json',
  headers: {
    'Content-Type': 'application/json',
  },
});

axios_.interceptors.request.use(
  config => {
    const { method, url, data } = config;
    console.log(`=> HTTP ${method?.toUpperCase()} ${url} ${data ? JSON.stringify(data) : ''}`.slice(0, 399));

    return config;
  },
  error => {
    throw error;
  },
);

axios_.interceptors.response.use(
  response => {
    console.log(`<= Response: ${response.status} ${JSON.stringify(response.data)}`.slice(0, 399));
    return response;
  },
  error => {
    console.error(`Response error --> ${error.message}`);
    throw error;
  },
);

export default axios_;
