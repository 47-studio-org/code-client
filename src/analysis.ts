import { chunk } from 'lodash';

import { collectBundleFiles, prepareBundleHashes, prepareFilePath, composeFilePayloads } from './files';
import parseGitUri from './gitUtils';
import {
  getFilters,
  createBundle,
  createGitBundle,
  extendBundle,
  uploadFiles,
  checkBundle,
  getAnalysis,
  AnalysisStatus,
  IResult,
  RemoteBundle,
  GetAnalysisResponseDto,
  AnalysisFailedResponse,
  AnalysisFinishedResponse,
} from './http';
import emitter from './emitter';
import { defaultBaseURL, MAX_PAYLOAD } from './constants';

import { ISupportedFiles, IFileInfo } from './interfaces/files.interface';
import { AnalysisSeverity, IAnalysisResult } from './interfaces/analysis-result.interface';

// 1. Create a bundle for paths from scratch. Return bundle info together with request details and analysis results. Create a class Bundle for this
// 2. class Bundle will implement method extend, that will conduct analysis and return another Bundle instance
// 3. Create a queue, that would manage bundle extensions

async function createRemoteBundle(
  baseURL: string,
  sessionToken: string,
  fileHashes: IFileInfo[],
  maxPayload = MAX_PAYLOAD,
): Promise<IResult<RemoteBundle> | null> {
  let response: IResult<RemoteBundle> | null = null;

  const fileChunks = chunk(fileHashes, maxPayload / 300);
  emitter.createBundleProgress(0, fileChunks.length);
  for (const [i, chunkedFiles] of fileChunks.entries()) {
    const files = Object.fromEntries(chunkedFiles.map(d => [prepareFilePath(d.path), d.hash]));

    if (response === null) {
      // eslint-disable-next-line no-await-in-loop
      response = await createBundle({
        baseURL,
        sessionToken,
        files,
      });
    } else {
      // eslint-disable-next-line no-await-in-loop
      response = await extendBundle({
        baseURL,
        sessionToken,
        bundleId: response.value.bundleId,
        files,
      });
    }

    emitter.createBundleProgress(i, fileChunks.length);

    if (response.type === 'error') {
      // TODO: process Error
      return response;
    }
  }

  return response;
}

/**
 * Splits files in buckets and upload in parallel
 * @param baseURL
 * @param sessionToken
 * @param remoteBundle
 */
export async function uploadRemoteBundle(
  baseURL: string,
  sessionToken: string,
  remoteBundle: RemoteBundle,
): Promise<boolean> {
  let uploadedFiles = 0;
  emitter.uploadBundleProgress(0, remoteBundle.missingFiles.length);

  const uploadFileChunks = async (bucketFiles: IFileInfo[]): Promise<boolean> => {
    const resp = await uploadFiles({
      baseURL,
      sessionToken,
      bundleId: remoteBundle.bundleId,
      content: bucketFiles.map(f => {
        return { fileHash: f.hash, fileContent: f.content || '' };
      }),
    });

    if (resp.type !== 'error') {
      uploadedFiles += bucketFiles.length;
      emitter.uploadBundleProgress(uploadedFiles, remoteBundle.missingFiles.length);
      return true;
    }

    return false;
  };

  const tasks = [];
  for (const bucketFiles of composeFilePayloads(remoteBundle.missingFiles)) {
    tasks.push(uploadFileChunks(bucketFiles));
  }

  if (tasks.length) {
    return (await Promise.all(tasks)).some(r => !!r);
  }
  return true;
}

async function pollAnalysis(
  baseURL: string,
  sessionToken: string,
  bundleId: string,
  useLinters: boolean,
  severity: AnalysisSeverity,
): Promise<IResult<AnalysisFailedResponse | AnalysisFinishedResponse>> {
  let analysisResponse: IResult<GetAnalysisResponseDto>;
  let analysisData: GetAnalysisResponseDto;

  emitter.analyseProgress({
    status: AnalysisStatus.fetching,
    progress: 0,
  });

  while (true) {
    // eslint-disable-next-line no-await-in-loop
    analysisResponse = await getAnalysis({
      baseURL,
      sessionToken,
      bundleId,
      useLinters,
      severity,
    });

    if (analysisResponse.type === 'error') {
      return analysisResponse;
    }

    analysisData = analysisResponse.value;

    if (
      analysisData.status === AnalysisStatus.fetching ||
      analysisData.status === AnalysisStatus.analyzing ||
      analysisData.status === AnalysisStatus.dcDone
    ) {
      // Report progress of fetching
      emitter.analyseProgress(analysisData);
    } else if (analysisData.status === AnalysisStatus.done) {
      // Return data of analysis
      return analysisResponse as IResult<AnalysisFinishedResponse>;
    } else if (analysisData.status === AnalysisStatus.failed) {
      // Report failure of analysing
      return analysisResponse as IResult<AnalysisFailedResponse>;
    }
  }
}

export async function analyzeFolders(
  baseURL = defaultBaseURL,
  sessionToken = '',
  includeLint = false,
  severity = AnalysisSeverity.info,
  paths: string[],
  maxPayload = MAX_PAYLOAD,
): Promise<FileBundle> {
  // Get supported filters and test baseURL for correctness and availability
  const resp = await getFilters(baseURL);
  if (resp.type === 'error') {
    throw new Error('baseURL is incorrect or server is not reachable now');
  }
  const supportedFiles = resp.value;

  // Scan directories and find all suitable files

  emitter.scanFilesProgress(0); // TODO: report progress on scanning too
  const bundleFiles = collectBundleFiles(paths, supportedFiles);
  emitter.scanFilesProgress(bundleFiles.length); // TODO: report progress on scanning too

  // Annotate all suitable files with meta information including hashe
  const fileHashes = [];
  let filesCount = 0;
  emitter.computeHashProgress(0, bundleFiles.length);
  for (const h of prepareBundleHashes(bundleFiles, maxPayload)) {
    fileHashes.push(h);
    filesCount += 1;
    emitter.computeHashProgress(filesCount, bundleFiles.length);
  }

  // Create remote bundle
  let bundleResponse = await createRemoteBundle(baseURL, sessionToken, fileHashes, maxPayload);
  if (bundleResponse === null) {
    throw new Error('File list is empty');
  } else if (bundleResponse.type === 'error') {
    throw bundleResponse.error;
  }

  // Fulfill remove bundle by uploading only missing files (splitted in chunks)
  // Check remove bundle to make sure no missing files left
  let remoteBundle = bundleResponse.value;
  if (remoteBundle.missingFiles.length) {
    const isUploaded = await uploadRemoteBundle(baseURL, sessionToken, remoteBundle);
    if (!isUploaded) {
      throw new Error('Failed to upload some files');
    }
    bundleResponse = await checkBundle({
      baseURL,
      sessionToken,
      bundleId: remoteBundle.bundleId,
    });
    if (bundleResponse.type === 'error') {
      throw new Error('Failed to get remote bundle');
    }
    remoteBundle = bundleResponse.value;
  }

  // Call remote bundle for analysis results and emit intermediate progress
  const analysisData = await pollAnalysis(baseURL, sessionToken, remoteBundle.bundleId, includeLint, severity);

  if (analysisData.type === 'error') {
    throw analysisData.error;
  } else if (analysisData.value.status === AnalysisStatus.failed) {
    throw new Error('Analysis has failed');
  }

  // Create bundle instance to handle extensions
  const bundle = new FileBundle(
    baseURL,
    sessionToken,
    includeLint,
    severity,
    remoteBundle.bundleId,
    analysisData.value.analysisResults,
    analysisData.value.analysisURL,
  );

  bundle.supportedFiles = supportedFiles;
  bundle.paths = paths;
  return bundle;
}

export async function analyzeGit(
  baseURL = defaultBaseURL,
  sessionToken = '',
  includeLint = false,
  severity = AnalysisSeverity.info,
  gitUri: string,
): Promise<GitBundle> {
  const repoKey = parseGitUri(gitUri);
  if (!repoKey) {
    throw new Error('Failed to parse git uri');
  }

  const bundleResponse = await createGitBundle({ baseURL, sessionToken, ...repoKey });
  if (bundleResponse.type === 'error') {
    throw new Error('Failed to find last commit hash');
  }
  const { bundleId } = bundleResponse.value;

  // Call remote bundle for analysis results and emit intermediate progress
  const analysisData = await pollAnalysis(baseURL, sessionToken, bundleId, includeLint, severity);

  if (analysisData.type === 'error') {
    throw analysisData.error;
  } else if (analysisData.value.status === AnalysisStatus.failed) {
    throw new Error('Analysis has failed');
  }

  // Create bundle instance to handle extensions
  const bundle = new GitBundle(
    baseURL,
    sessionToken,
    includeLint,
    severity,
    bundleId,
    analysisData.value.analysisResults,
    analysisData.value.analysisURL,
  );
  bundle.gitUri = gitUri;
  return bundle;
}

class BundleBase {
  private readonly baseURL: string;
  private readonly sessionToken: string;
  private readonly includeLint: boolean;
  private readonly severity: AnalysisSeverity;
  public readonly bundleId: string;
  public readonly analysisResults: IAnalysisResult;
  public readonly analysisUrl: string;

  constructor(
    baseURL: string,
    sessionToken: string,
    includeLint: boolean,
    severity: AnalysisSeverity,
    bundleId: string,
    analysisResults: IAnalysisResult,
    analysisUrl: string,
  ) {
    this.baseURL = baseURL;
    this.sessionToken = sessionToken;
    this.includeLint = includeLint;
    this.severity = severity;
    this.bundleId = bundleId;
    this.analysisResults = analysisResults;
    this.analysisUrl = analysisUrl;
  }
}

class GitBundle extends BundleBase {
  public gitUri: string;
}

class FileBundle extends BundleBase {
  public paths: string[];
  public supportedFiles: ISupportedFiles;

  // public async extend(files: string[], removedFiles: string[]): Promise<FileBundle> {
  //   return this;
  // }

  //     await startAnalysisLoop({ baseURL, sessionToken, bundleId: this.bundleId }).catch(error => {
  //       emitter.sendError(error);
  //       throw error;
  //     });
  //   } catch (error) {
  //     emitter.sendError(error);
  //     throw error;
  //   }
  // }
}
