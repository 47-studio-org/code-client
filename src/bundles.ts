/* eslint-disable no-await-in-loop */
import chunk from 'lodash.chunk';
import pick from 'lodash.pick';

import { BundleFiles, FileInfo } from './interfaces/files.interface';

import { composeFilePayloads, resolveBundleFiles } from './files';
import {
  CreateBundleErrorCodes,
  CheckBundleErrorCodes,
  ExtendBundleErrorCodes,
  createBundle,
  extendBundle,
  checkBundle,
  Result,
  RemoteBundle,
} from './http';
import { MAX_PAYLOAD, MAX_UPLOAD_ATTEMPTS } from './constants';
import emitter from './emitter';

type BundleErrorCodes = CreateBundleErrorCodes | CheckBundleErrorCodes | ExtendBundleErrorCodes;

async function* prepareRemoteBundle(
  baseURL: string,
  sessionToken: string,
  files: FileInfo[],
  removedFiles: string[] = [],
  existingBundleHash: string | null = null,
  maxPayload = MAX_PAYLOAD,
  source: string,
): AsyncGenerator<Result<RemoteBundle, BundleErrorCodes>> {
  let response: Result<RemoteBundle, BundleErrorCodes>;
  let bundleHash = existingBundleHash;

  const fileChunks = chunk(files, maxPayload / 300);
  emitter.createBundleProgress(0, fileChunks.length);
  for (const [i, chunkedFiles] of fileChunks.entries()) {
    const paramFiles = chunkedFiles.reduce((d, f) => ({ ...d, [f.bundlePath]: f.hash }), {} as BundleFiles);

    if (bundleHash === null) {
      // eslint-disable-next-line no-await-in-loop
      response = await createBundle({
        baseURL,
        sessionToken,
        files: paramFiles,
        source,
      });
    } else {
      // eslint-disable-next-line no-await-in-loop
      response = await extendBundle({
        baseURL,
        sessionToken,
        bundleHash,
        files: paramFiles,
        removedFiles,
      });
    }

    emitter.createBundleProgress(i + 1, fileChunks.length);

    if (response.type === 'error') {
      // TODO: process Error
      yield response;
      break;
    }
    bundleHash = response.value.bundleHash;

    yield response;
  }
}

/**
 * Splits files in buckets and upload in parallel
 * @param baseURL
 * @param sessionToken
 * @param remoteBundle
 */
export async function uploadRemoteBundle(
  baseURL: string,
  sessionToken: string,
  bundleHash: string,
  files: FileInfo[],
  maxPayload = MAX_PAYLOAD,
): Promise<boolean> {
  let uploadedFiles = 0;
  emitter.uploadBundleProgress(0, files.length);

  const uploadFileChunks = async (bucketFiles: FileInfo[]): Promise<boolean> => {
    const resp = await extendBundle({
      baseURL,
      sessionToken,
      bundleHash,
      files: bucketFiles.reduce((d, f) => ({ ...d, [f.bundlePath]: pick(f, ['hash', 'content']) }), {}),
    });

    if (resp.type !== 'error') {
      uploadedFiles += bucketFiles.length;
      emitter.uploadBundleProgress(uploadedFiles, files.length);
      return true;
    }

    return false;
  };

  const tasks = [];
  for (const bucketFiles of composeFilePayloads(files, maxPayload)) {
    tasks.push(uploadFileChunks(bucketFiles));
  }

  if (tasks.length) {
    return (await Promise.all(tasks)).some(r => !!r);
  }
  return true;
}

async function fullfillRemoteBundle(
  baseURL: string,
  sessionToken: string,
  baseDir: string,
  remoteBundle: RemoteBundle,
  maxPayload = MAX_PAYLOAD,
  maxAttempts = MAX_UPLOAD_ATTEMPTS,
): Promise<RemoteBundle> {
  // Fulfill remote bundle by uploading only missing files (splitted in chunks)
  // Check remove bundle to make sure no missing files left
  let attempts = 0;
  while (remoteBundle.missingFiles.length && attempts < maxAttempts) {
    const missingFiles = await resolveBundleFiles(baseDir, remoteBundle.missingFiles);
    const isUploaded = await uploadRemoteBundle(
      baseURL,
      sessionToken,
      remoteBundle.bundleHash,
      missingFiles,
      maxPayload,
    );
    if (!isUploaded) {
      throw new Error('Failed to upload some files');
    }
    const bundleResponse = await checkBundle({
      baseURL,
      sessionToken,
      bundleHash: remoteBundle.bundleHash,
    });
    if (bundleResponse.type === 'error') {
      throw new Error('Failed to get remote bundle');
    }
    // eslint-disable-next-line no-param-reassign
    remoteBundle = bundleResponse.value;
    attempts += 1;
  }
  return remoteBundle;
}

export async function remoteBundleFactory(
  baseURL: string,
  sessionToken: string,
  files: FileInfo[],
  removedFiles: string[] = [],
  baseDir: string,
  existingBundleHash: string | null = null,
  maxPayload = MAX_PAYLOAD,
  source: string,
): Promise<RemoteBundle | null> {
  const bundleFactory = prepareRemoteBundle(
    baseURL,
    sessionToken,
    files,
    removedFiles,
    existingBundleHash,
    maxPayload,
    source,
  );
  let remoteBundle: RemoteBundle | null = null;

  for await (const response of bundleFactory) {
    if (response.type === 'error') {
      throw response.error;
    }

    remoteBundle = await fullfillRemoteBundle(baseURL, sessionToken, baseDir, response.value, maxPayload);
    if (remoteBundle.missingFiles.length) {
      throw new Error(`Failed to upload # files: ${remoteBundle.missingFiles.length}`);
    }
  }

  return remoteBundle;
}
