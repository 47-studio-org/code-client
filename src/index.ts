import { analyzeFolders, analyzeGit } from './analysis';
import emitter from './emitter';
import { startSession, checkSession, reportEvent, reportError } from './http';
import * as constants from './constants';

import { IFileBundle } from './interfaces/analysis-result.interface';

export {
  analyzeFolders,
  analyzeGit,
  startSession,
  checkSession,
  reportEvent,
  reportError,
  emitter,
  constants,
  IFileBundle,
};
