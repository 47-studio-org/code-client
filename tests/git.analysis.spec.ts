import * as fs from 'fs';

import { baseURL, sessionToken, TEST_TIMEOUT } from './constants/base';
import { analyzeGit } from '../src/analysis';
import jsonschema from 'jsonschema';
import { Log } from 'sarif';
import * as sarifSchema from './sarif-schema-2.1.0.json';
import { ErrorCodes } from '../src/constants';
import { IGitBundle } from '../src/interfaces/analysis-result.interface';
import axios from '../src/axios';

const oAuthToken = process.env.SNYK_OAUTH_KEY || '';
const sessionTokenNoRepoAccess = process.env.SNYK_API_KEY_NO_ACCESS || '';

// This trick is for automatic tests, where real oauth token is not available
const itif = (condition: boolean) => (condition ? it : it.skip);

describe('Functional test of analysis', () => {
  it(
    'analyze remote git without oid',
    async () => {
      const bundle = await analyzeGit({
        baseURL,
        sessionToken,
        includeLint: false,
        severity: 1,
        gitUri: 'git@github.com:DeepCodeAI/cli.git',
      });
      expect(bundle.analysisResults.files).toBeTruthy();
      expect(bundle.analysisResults.suggestions).toBeTruthy();
    },
    TEST_TIMEOUT,
  );

  itif(!!oAuthToken)(
    'analyze remote git with oid',
    async () => {
      const bundle = await analyzeGit({
        baseURL,
        sessionToken,
        includeLint: false,
        severity: 1,
        gitUri: 'git@github.com:DeepCodeAI/cli.git@320d98a6896f5376efe6cefefb6e70b46b97d566',
      });
      expect(bundle.analysisResults.files).toBeTruthy();
      expect(Object.keys(bundle.analysisResults.files).length).toEqual(1);
      expect(bundle.analysisResults.suggestions).toBeTruthy();
      expect(Object.keys(bundle.analysisResults.suggestions).length).toEqual(1);
    },
    TEST_TIMEOUT,
  );

  itif(!!(oAuthToken && sessionTokenNoRepoAccess))(
    'analyze private remote git with oAuthToken',
    async () => {
      let failedGit: IGitBundle | undefined;
      try {
        failedGit = await analyzeGit({
          baseURL,
          sessionToken: sessionTokenNoRepoAccess,
          includeLint: false,
          severity: 1,
          gitUri: 'git@github.com:DeepCodeAI/testcrepo.git',
        });
      } catch (failed) {
        expect(failed.statusCode).toEqual(ErrorCodes.unauthorizedBundleAccess);
      }
      expect(failedGit).toBe(undefined);

      const bundle = await analyzeGit({
        baseURL,
        sessionToken: sessionTokenNoRepoAccess,
        includeLint: false,
        severity: 1,
        gitUri: 'git@github.com:DeepCodeAI/testcrepo.git',
        sarif: false,
        oAuthToken,
      });
      expect(bundle.analysisResults.files).toBeTruthy();
      expect(bundle.analysisResults.suggestions).toBeTruthy();
    },
    TEST_TIMEOUT,
  );

  it(
    'CWE fields in analysis results',
    async () => {
      const bundle = await analyzeGit({
        baseURL,
        sessionToken,
        includeLint: false,
        severity: 1,
        gitUri: 'git@github.com:eclipse/che.git@75889e8c33601e8986e75bad74456cff39e511c0',
        sarif: true,
      });

      // Test DC JSON format first
      expect(Object.keys(bundle.analysisResults.suggestions).length).toEqual(134);

      const cweSuggestion = Object.values(bundle.analysisResults.suggestions).find(
        s => s.id === 'java%2Fdc_interfile_project%2FPT',
      );

      expect(cweSuggestion?.cwe).toEqual(['CWE-23']);
      expect(cweSuggestion?.title).toBeTruthy();
      expect(cweSuggestion?.text).toBeTruthy();

      expect(bundle.sarifResults?.runs[0].results?.length).toEqual(442);
      expect(bundle.sarifResults?.runs[0].tool?.driver.rules?.length).toEqual(134);

      const cweRule = bundle.sarifResults?.runs[0].tool?.driver.rules?.find(r => r.id === 'java/PT');
      expect(cweRule?.properties?.cwe).toContain('CWE-23');
      expect(cweRule?.shortDescription?.text).toEqual('Path Traversal');
    },
    TEST_TIMEOUT,
  );

  describe('detailed sarif tests', () => {
    let sarifResults: Log | undefined;
    it(
      'analyze remote git with oid and return sarif',
      async () => {
        const bundle = await analyzeGit({
          baseURL,
          sessionToken,
          includeLint: false,
          severity: 1,
          gitUri: 'git@github.com:DeepCodeAI/cli.git@320d98a6896f5376efe6cefefb6e70b46b97d566',
          sarif: true,
        });
        sarifResults = bundle.sarifResults;
        expect(!!bundle.sarifResults).toEqual(true);
      },
      TEST_TIMEOUT,
    );

    it(
      'analyze remote git and formatter sarif with zero supported files',
      async () => {
        const bundle = await analyzeGit({
          baseURL,
          sessionToken,
          includeLint: false,
          severity: 1,
          gitUri: 'git@github.com:DeepCodeAI/test-bigfiles.git@e7633ef98fba3ddc24e5bea27ae58d5b08b2f949',
          sarif: true,
        });
        expect(bundle.sarifResults?.runs[0].properties?.coverage).toEqual([
          {
            files: 3,
            isSupported: false,
            lang: 'Text',
          },
          {
            files: 1,
            isSupported: false,
            lang: 'Markdown',
          },
        ]);
      },
      TEST_TIMEOUT,
    );

    it(
      'analyze remote git and formatter sarif with supported files',
      async () => {
        const bundle = await analyzeGit({
          baseURL,
          sessionToken,
          includeLint: false,
          severity: 1,
          gitUri: 'git@github.com:OpenRefine/OpenRefine.git@437dc4d74110ce006b9f829fe05f461cc8ed1170',
          sarif: true,
        });
        expect(bundle.sarifResults?.runs[0].properties?.coverage).toMatchSnapshot();

        const numOfIssues = getNumOfIssues(bundle);
        const numOfIssuesInSarif = bundle.sarifResults?.runs[0].results?.length;

        expect(numOfIssuesInSarif).toEqual(numOfIssues);
      },
      TEST_TIMEOUT,
    );

    it('should match sarif schema', () => {
      const validationResult = jsonschema.validate(sarifResults, sarifSchema);
      // this is to debug any errors found
      // const json = JSON.stringify(validationResult)
      // fs.writeFile('sarif_validation_log.json', json, 'utf8', ()=>null);
      expect(validationResult.errors.length).toEqual(0);
    });
  });
});

describe('Custom request options', () => {
  beforeAll(() => {
    jest.mock('axios');
    axios.request = jest.fn().mockRejectedValue({});
  });

  it(
    'passes custom options correctly',
    async () => {
      try {
        await analyzeGit(
          {
            baseURL,
            sessionToken,
            includeLint: false,
            severity: 1,
            gitUri: 'git@github.com:DeepCodeAI/cli.git',
          },
          { headers: { 'X-test-header': 'Snyk' } },
        );
      } catch (e) {
        // expected to fail, we are interested in correct propagation of headers only
      }
      expect((axios.request as jest.Mock).mock.calls[0][0]).toMatchObject({ headers: { 'X-test-header': 'Snyk' } });
    },
    TEST_TIMEOUT,
  );

  afterAll(() => {
    jest.resetAllMocks();
  });
});

function getNumOfIssues(bundle: IGitBundle): number {
  let numberOfIssues = 0;

  Object.keys(bundle.analysisResults.files).forEach(key =>
    Object.keys(bundle.analysisResults.files[key]).forEach(
      issueId => (numberOfIssues += bundle.analysisResults.files[key][issueId].length),
    ),
  );

  return numberOfIssues;
}
