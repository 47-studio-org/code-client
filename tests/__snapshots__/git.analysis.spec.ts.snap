// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Functional test of analysis detailed sarif tests analyze remote git and formatter sarif with supported files 1`] = `
Array [
  Object {
    "files": 115,
    "isSupported": false,
    "lang": "Unknown",
  },
  Object {
    "files": 34,
    "isSupported": false,
    "lang": "Markdown",
  },
  Object {
    "files": 6,
    "isSupported": false,
    "lang": "YAML",
  },
  Object {
    "files": 3,
    "isSupported": false,
    "lang": "Shell",
  },
  Object {
    "files": 39,
    "isSupported": false,
    "lang": "Text",
  },
  Object {
    "files": 34,
    "isSupported": false,
    "lang": "XML",
  },
  Object {
    "files": 4,
    "isSupported": false,
    "lang": "INI",
  },
  Object {
    "files": 122,
    "isSupported": true,
    "lang": "JavaScript",
  },
  Object {
    "files": 81,
    "isSupported": false,
    "lang": "JSON",
  },
  Object {
    "files": 13,
    "isSupported": false,
    "lang": "CSS",
  },
  Object {
    "files": 21,
    "isSupported": false,
    "lang": "Java Properties",
  },
  Object {
    "files": 6,
    "isSupported": false,
    "lang": "SVG",
  },
  Object {
    "files": 78,
    "isSupported": true,
    "lang": "HTML",
  },
  Object {
    "files": 45,
    "isSupported": false,
    "lang": "Less",
  },
  Object {
    "files": 989,
    "isSupported": true,
    "lang": "Java",
  },
  Object {
    "files": 4,
    "isSupported": false,
    "lang": "SQLPL",
  },
  Object {
    "files": 8,
    "isSupported": false,
    "lang": "CSV",
  },
  Object {
    "files": 1,
    "isSupported": false,
    "lang": "Batchfile",
  },
]
`;
